### PR TITLE
Fuse coefficient loops

### PR DESCRIPTION
**Pull request by IgorBaratta**
Current branch:
``` c++
    for (int ic = 0; ic < 56; ++ic)
    {
        _w_0_0[ic] = w[ic * 3];
        _w_0_1[ic] = w[ic * 3 + 1];
        _w_0_2[ic] = w[ic * 3 + 2];
    }
    for (int iq = 0; iq < 45; ++iq)
    {
        // Quadrature loop body setup for quadrature rule a86
        // Varying computations for quadrature rule a86
        double w0_d100_c0 = 0.0;
        double w0_d010_c0 = 0.0;
        double w0_d001_c0 = 0.0;
        double w0_d100_c1 = 0.0;
        double w0_d010_c1 = 0.0;
        double w0_d001_c1 = 0.0;
        double w0_d100_c2 = 0.0;
        double w0_d010_c2 = 0.0;
        double w0_d001_c2 = 0.0;
        for (int ic = 0; ic < 56; ++ic)
        {
            w0_d100_c0 += _w_0_0[ic] * FE18_C0_D100_Qa86[0][0][iq][ic];
            w0_d010_c0 += _w_0_0[ic] * FE18_C0_D010_Qa86[0][0][iq][ic];
            w0_d001_c0 += _w_0_0[ic] * FE18_C0_D001_Qa86[0][0][iq][ic];
            w0_d100_c1 += _w_0_1[ic] * FE18_C0_D100_Qa86[0][0][iq][ic];
            w0_d010_c1 += _w_0_1[ic] * FE18_C0_D010_Qa86[0][0][iq][ic];
            w0_d001_c1 += _w_0_1[ic] * FE18_C0_D001_Qa86[0][0][iq][ic];
            w0_d100_c2 += _w_0_2[ic] * FE18_C0_D100_Qa86[0][0][iq][ic];
            w0_d010_c2 += _w_0_2[ic] * FE18_C0_D010_Qa86[0][0][iq][ic];
            w0_d001_c2 += _w_0_2[ic] * FE18_C0_D001_Qa86[0][0][iq][ic];
        }
```

Main branch:
```c++
    // Auxiliary array to enable unit-stride access in coefficient computations.
    double _w_0_0[56];
    for (int ic = 0; ic < 56; ++ic)
        _w_0_0[ic] = w[ic * 3];
    // Auxiliary array to enable unit-stride access in coefficient computations.
    double _w_0_1[56];
    for (int ic = 0; ic < 56; ++ic)
        _w_0_1[ic] = w[ic * 3 + 1];
    // Auxiliary array to enable unit-stride access in coefficient computations.
    double _w_0_2[56];
    for (int ic = 0; ic < 56; ++ic)
        _w_0_2[ic] = w[ic * 3 + 2];
    for (int iq = 0; iq < 45; ++iq)
    {
        // Quadrature loop body setup for quadrature rule a86
        // Varying computations for quadrature rule a86
        double w0_d100_c0 = 0.0;
        for (int ic = 0; ic < 56; ++ic)
            w0_d100_c0 += _w_0_0[ic] * FE18_C0_D100_Qa86[0][0][iq][ic];
        double w0_d010_c0 = 0.0;
        for (int ic = 0; ic < 56; ++ic)
            w0_d010_c0 += _w_0_0[ic] * FE18_C0_D010_Qa86[0][0][iq][ic];
        double w0_d001_c0 = 0.0;
        for (int ic = 0; ic < 56; ++ic)
            w0_d001_c0 += _w_0_0[ic] * FE18_C0_D001_Qa86[0][0][iq][ic];
        double w0_d100_c1 = 0.0;
        for (int ic = 0; ic < 56; ++ic)
            w0_d100_c1 += _w_0_1[ic] * FE18_C0_D100_Qa86[0][0][iq][ic];
        double w0_d010_c1 = 0.0;
        for (int ic = 0; ic < 56; ++ic)
            w0_d010_c1 += _w_0_1[ic] * FE18_C0_D010_Qa86[0][0][iq][ic];
        double w0_d001_c1 = 0.0;
        for (int ic = 0; ic < 56; ++ic)
            w0_d001_c1 += _w_0_1[ic] * FE18_C0_D001_Qa86[0][0][iq][ic];
        double w0_d100_c2 = 0.0;
        for (int ic = 0; ic < 56; ++ic)
            w0_d100_c2 += _w_0_2[ic] * FE18_C0_D100_Qa86[0][0][iq][ic];
        double w0_d010_c2 = 0.0;
        for (int ic = 0; ic < 56; ++ic)
            w0_d010_c2 += _w_0_2[ic] * FE18_C0_D010_Qa86[0][0][iq][ic];
        double w0_d001_c2 = 0.0;
        for (int ic = 0; ic < 56; ++ic)
            w0_d001_c2 += _w_0_2[ic] * FE18_C0_D001_Qa86[0][0][iq][ic];
```

### CODE REVIEW

File: ffcx/codegeneration/integrals.py [hunk 1]
```diff
@@ -691,3 +688,29 @@ def generate_block_parts(self, quadrature_rule: QuadratureRule, blockmap: Tuple,
...
+        Ideally, we should define a cost function to determine how many loops should fuse at a time.
+        """
+        L = self.backend.language
+
+        loops = collections.defaultdict(list)
+        pre_loop = []
+        for access, definition in definitions.items():
+            for d in definition:
+                if isinstance(d, L.ForRange):
+                    loops[(d.begin, d.end)] += [d.body]
```
Comment (michalhabera) [inline ffcx/codegeneration/integrals.py:707]: I think you should add ForRange's iteration symbol (e.g. "ic") to dicts keys. The function looks general and could accidentally fuse across different loops if used elsewhere.

File: ffcx/codegeneration/integrals.py [hunk 2]
```diff
@@ -691,3 +688,29 @@ def generate_block_parts(self, quadrature_rule: QuadratureRule, blockmap: Tuple,
         quadparts += [pre_loop, hoist, body]
 
         return preparts, quadparts
+
+    def fuse_loops(self, definitions):
+        """
+        Merge a sequence of loops with the same iteration space into a single loop.
+        Loop fusion improves data locality, cache reuse and decreases the loop control overhead.
+        NOTE: Loop fusion might increase the pressure on register allocation.
+        Ideally, we should define a cost function to determine how many loops should fuse at a time.
```
Comment (michalhabera) [inline ffcx/codegeneration/integrals.py:698]: Yep, but I am not a big fan of doing these compiler-alike optimisations in FFCx codegen layer. Would be much cleaner to have a separate "compiler pass" for these.

Comment (IgorBaratta) [reply ffcx/codegeneration/integrals.py:698]: I agree that it's not ideal. But we would lose "locality". 
Parsing the entire code is much more complicated than applying optimizations to, say, coefficient computations.

File: ffcx/codegeneration/integrals.py (hunk 1 above)
Comment (IgorBaratta) [reply ffcx/codegeneration/integrals.py:707]: Added the index to the dict key:
```python 
loops[(d.index, d.begin, d.end)] += [d.body]
```

**Review by garth-wells — APPROVED** [review]

**Pull request merged**